text/0002-interfaces.md

```diff
@@ -0,0 +1,300 @@
+  * `signature.__getitem__(path)` looks up a member by its path. The flow of the member is flipped as appropriate.
+  * `signature.flip()` returns a signature where every member is `member.flip()`ped.
+  * `signature.compatible(object)` checks whether an arbitrary Python object is compatible with this signature. To be compatible with a signature:
+    - for every member of the signature, the object must have a corresponding attribute
+    - if the member is a port, the attribute value must be a value-castable whose `as_value()` method returns a `Signal` that has the same width, signedness, and reset value as the member, and is not reset-less
+      - a warning may be emitted if the `.shape` of the member and of the attribute are different
+    - if the member is an interface, the attribute value must be compatible with the signature of the member
+  * `signature.__call__()` freezes this signature and creates an interface from it. This method is expected to be routinely overridden in `Signature` subclasses to perform actions specific to a particular signature. The interface is created as follows:
+    1. A fresh Python object `obj` is created using `object()`
+    2. `obj.signature` is set to `signature`
+    3. For every member of the signature, an attribute with the member's name is set on `obj`, where the value is:
+      * If the member is a port, `Signal(member.shape, reset=member.reset)`.
+      * If the member is a signature, `member.signature.__call__()`.
+
+
+### Interface connection
+
+Interfaces may be connected to each other using the `amaranth.lib.module.connect(m, first, *rest)` free function.
+
+This function connects interfaces that satisfy the following requirements:
+* For every port member with a given `path` and `first_shape` in `first`, there is a port member with the same `path` in each of `rest` with shape `rest_shape`, where `Shape.cast(first_shape) == Shape.cast(rest_shape)`, and there are no other members in any of `rest`. In other words, the width and signedness of all of port members must be equal, but the shape-castable objects specifying the width and signedness do not have to be.
+  * It is expected that any mismatch in signatures will be resolved (through wrappers, mutating signature members, or otherwise) before interfaces are being connected.
+* Either:
+  1. There is exactly one interface in `rest`, and for every pair of members `first_member` and `rest_member`, `first_member.flow == ~rest_member.flow`.
+  2. There is any amount of interfaces in `rest`, and for every pair of members `first_member` and `rest_member`, `first_member.flow == Out` and `rest_member.flow == In`.
+
+In both of the cases (1) and (2), the function, for each pair of an input port and an output of port with the same path, the function connects these as follows:
+```python
+m.d.comb += output_port.eq(input_port)
```

Comment:
I think the overloading of "out"/"output" and "in"/"input" in this illustrative code block has great potential to confuse — e.g. it's the _output stream_ of `SequenceSource` (which is described in the source with the member `Out[16]`) from the above example which would be the `input_port` here, and it's the _input stream_ of the `NumberSink` (described with `In[16]` in its signature) which is the `output_port`.

In the context of knowing only fan-out is supported (as noted later in § Future work), this makes sense, but in the context of case (2) saying "if you have multiple `rest` interfaces, all their members must be `Flow.In`", it may read as if there would be multiple `input_port`s and thus some kind of fan-in instead.